packages/dds/shared-object-base/src/serializer.ts

```diff
@@ -58,7 +59,103 @@ export interface IFluidSerializer {
+	| { readonly [P in string]?: FluidSerializableReadOnly };
+
+/**
+ * Check if a value in {@link FluidSerializableReadOnly} data is an {@link @fluidframework/core-interfaces#IFluidHandle}.
+ *
+ * Warning: Non-IFluidHandle objects with a key "IFluidHandle" will cause an assert.
+ * Prior to version 2.0.0-internal.7.4.0, they were unconditionally interpreted as IFluidHandle.
+ *
+ * @remarks
+ * Its possible to have serializable data that looks like an {@link @fluidframework/core-interfaces#IFluidHandle} but isn't one.
+ * This function cannot distinguish these from real IFluidHandles in all cases: changes to the IFluidHandle would be required to ensure this is always possible.
+ *
+ * For now this is a best effort implementation.
+ *
+ * @privateRemarks
+ * Any of the following changes to IFluidHandle would solve the ambiguity problem:
+ *
+ * - The "IFluidHandle" property could be required to be a cyclic reference back to the parent.
+ *
+ * - IFluidHandle could be made into a class which could be detected with `instanceof`.
+ *
+ * - IFluidHandle could be identified using a dedicated symbol.
+ *
+ * @public
+ */
+export function isFluidHandle(value: undefined | FluidSerializableReadOnly): value is IFluidHandle {
+	if (typeof value !== "object" || value === null) {
+		return false;
+	}
+	const partialHandle = value as Partial<IFluidHandle>;
+	const innerHandle = partialHandle.IFluidHandle;
+	if (innerHandle === undefined) {
+		return false;
+	}
+
+	// At this point, the value is known to be an object with an IFluidHandle.
+	// This function will heuristically validate that object really is an IFluidHandle.
+	// If the object fails this check, it will assert to detect problematic data that would have incorrectly been treated as a handle in previous versions of the serializer.
+
+	// Since json compatible data shouldn't have methods, and IFluidHandle requires one, use that as a redundant check:
+	const getMember = (value as Partial<IFluidHandle>).get;
+	assert(
+		typeof getMember === "function",
+		"Fluid handle detection found IFluidHandle field, but not have a get method",
+	);
+
+	// If the handle is not a cyclic reference, validate the inner one as well:
+	if (value !== innerHandle) {
+		assert(
+			typeof innerHandle === "object" && innerHandle !== null,
+			"Fluid handle detection found IFluidHandle field, but it was not an object",
+		);
+
+		assert(
+			innerHandle.IFluidHandle?.IFluidHandle?.IFluidHandle?.IFluidHandle !== undefined,
+			"Fluid handle detection found IFluidHandle field, but it did not have IFluidHandle members recursively",
+		);
+
+		const innerGetMember = (innerHandle as Partial<IFluidHandle>).get;
+		assert(
+			typeof innerGetMember === "function",
+			"Fluid handle detection found IFluidHandle field, but it did not have a get method",
+		);
+	}
+
+	return true;
+}
+
+/**
+ * Data Store serializer implementation.
+ *
+ * @privateRemarks
+ * Since this type is package exported (not just the Interface above),
```

Comment:
It's `@internal` now!  I had wanted to stop it from being public anyway, and it just happened in bulk.  Maybe it's kind to wait until after internal.8.0 "just in case", but we should be good to change stuff like this with more liberty now AFAIK.